policy/fetcher.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright (C)  2018 Yahoo Japan Corporation Athenz team.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/kpango/fastime"
+	"github.com/kpango/glg"
+	"github.com/pkg/errors"
+)
+
+// SignedPolicyVerifier type defines the function signature to verify a signed policy.
+type SignedPolicyVerifier func(*SignedPolicy) error
+
+// Fetcher represents fetcher object for fetching signed policy
+type Fetcher interface {
+	Domain() string
+	Fetch(context.Context) (*SignedPolicy, error)
+	FetchWithRetry(context.Context) (*SignedPolicy, error)
+}
+
+type fetcher struct {
+
+	// etag related
+	expireMargin time.Duration
+
+	// retry related
+	retryInterval time.Duration
+	retryMaxCount int
+
+	// athenz related
+	domain     string
+	athenzURL  string
+	spVerifier SignedPolicyVerifier
+
+	client      *http.Client
+	policyCache unsafe.Pointer
```

Comment:
we should use *atomic.Value instead of unsafe.Pointer here

Reply:
they are the same?

unsafe seems faster. (only load is tested...)
https://blog.dubbelboer.com/2015/08/23/rwmutex-vs-atomicvalue-vs-unsafepointer.html